Create Timeframe component

bernardodiasc opened the pull request:
**Release Type:** *Non-breaking changes*

Fixes https://x-team-internal.atlassian.net/browse/XP-3701

## Description

Create Timeframe component.

## Checklist

- [x] set yourself as an assignee
- [x] set appropriate labels for a PR (`In Review` or `In Progress` depending on its status)
- [x] move respective JIRA issue to the `PEER REVIEW` column
- [x] Flow typechecks passed (`npm run flow`)
- [x] if added a new UI component that is meant to be used directly by other apps, remember to export that from the appropriate bundle file: `src/index.js` and/or `src/registration.js`
- [x] component's documentation (`.stories.js` file) is changed or added accordingly to reflect any new or updated use cases and/or debug cases
- [x] **manually tested the app** by running it in several different browsers (Firefox, Chrome, Opera, Safari, MS IE/Edge, etc.) and checked nothing is broken and operates as expected!

## Steps to Test or Reproduce

- Check the showcase in Storybook http://localhost:9001/?path=/story/core-components-form-components-timeframe--basic-usage

Please take a look at the other Debug showcases, it has some edge cases.

For instance, the start date should never be bigger than an end date. So: 

- If the **start year** is the same of the **end year**, then only the available months to select in the **end month** will be bigger or equal than the **start month**
- If the **start year** is the smaller of the **end year**, all **end month** with be available
- The **end year** options will only have years equal or bigger than **start year**
- If the `Still working there` is checked, then the value returned will ignore the end date (as end date will be undefined)

## Impacted Areas in Application

New `Timeframe` component was added.

## Screenshots

![2020-04-24 12 13 37](https://user-images.githubusercontent.com/131859/80228207-215ee300-8625-11ea-8440-ea40e33ab782.gif)

jaspax commented:
I tried to use this component and noticed a few things that were off:

* The year selector is sorted in inverse order with the highest (ie most recent) years at the top. I can see why you might think this is a good idea, but it's contrary to my expectations and caused confusion. Can we get a double-check whether this is a good idea?
* When clicking, "Still working there", can we clear out the values for the field with blanks for dashes?

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+}
+
+type Props = {
+  startDate?: string,
```
Shouldn't `startDate` be required? We never want startDate to be null.

bernardodiasc replied:
It's null when the user haven't added any start date yet.

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+
+type Props = {
+  startDate?: string,
+  endDate?: string,
```
More importantly, why are these props STRINGS?! They really ought to be `Date` unless there's some really good reason that I'm not seeing.

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+    const startMonth = Number(startDate.slice(5, 7))
+    const startYear = Number(startDate.slice(0, 4))
+    let endMonth = Number(endDate.slice(5, 7))
+    let endYear = Number(endDate.slice(0, 4))
```
Slicing date strings to get values seems extremely fragile. Please use a `Date`.

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+  startDate?: string,
+  endDate?: string,
+  noEndDate?: boolean,
+  onChange?: ({ startDate: string, endDate: ?string }) => void
```
The `onChange` event should likewise take `Date` arguments rather than strings

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+type Props = {
+  startDate?: string,
+  endDate?: string,
+  noEndDate?: boolean,
```
In the `onChange` handler we're handling the "no end date" case simply by passing a null `endDate`. Why not do this here? Why do we need this property at all?

bernardodiasc replied:
The end date can be undefined before the user add a value and that doesn't mean the checkbox that disable the field should be checked.

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+    }, this.validateValues)
+  }
+
+  validateValues = () => {
```
Why is this called `validateValues`? I don't see any validation going on here. Since method seems to do a simple call to `setState` and then set the callback, and I'm not sure why it couldn't just be unified with `handleOnChange`.

bernardodiasc replied:
This method handles cases where end dates are previous than start dates. Because there are 4 fields and the user can simply set these in any order, we need to validate those situations. I provided some explanation regarding these situations in the demo https://youtu.be/I6qvFBPE2K4, please let me know if you think something should be different from that, thanks!

jaspax replied:
Okay, I had to re-read the function to see how this method actually accomplishes that, because it's not clear from the names how `this.getEndMonthsList` is actually validating anything. Maybe add something to the name to make the semantics clearer.

jaspax replied:
I mean the functions `getEndMonthsList`. Maybe something like `getValidEndMonths`

jaspax commented on src/components/ui/Timeframe.js:
```diff
@@ -0,0 +1,283 @@
+    const { startMonth, startYear, endMonth, endYear } = this.state
+    this.setState({
+      startMonth: startMonth && this.getOptionValue(MONTHS, startMonth.value),
+      startYear: startYear && this.getOptionValue(YEARS, startYear.value),
```
The checks for `startMonth` and `startYear` are completely redundant as far as I can tell.

bernardodiasc commented:
hey @jaspax, thanks for the feedback. I addressed your suggestions. Given the discussion in the dev-design collab channel, here's how the component looks like now:

![2020-04-27 13 58 11](https://user-images.githubusercontent.com/131859/80402598-7f453200-8894-11ea-826d-53775f2629db.gif)
![2020-04-27 13 58 55](https://user-images.githubusercontent.com/131859/80402605-810ef580-8894-11ea-9ae6-e60c3bb66e3d.gif)
